src/cascadia/TerminalSettingsModel/MatchProfilesEntry.cpp

```diff
@@ -36,41 +36,71 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+
         JsonUtils::GetValueForKey(json, SourceKey, entry->_Source);
+        entry->_validateSource();
 
         return entry;
     }
 
+    // Returns true if all regexes are valid, false otherwise
+    bool MatchProfilesEntry::ValidateRegexes() const
+    {
+        return !(_invalidName || _invalidCommandline || _invalidSource);
+    }
+
+#define DEFINE_VALIDATE_FUNCTION(name)                               \
+    void MatchProfilesEntry::_validate##name() noexcept              \
+    {                                                                \
+        _invalid##name = false;                                      \
+        if (_##name.empty())                                         \
+        {                                                            \
+            /* empty field is valid*/                                \
+            return;                                                  \
+        }                                                            \
+        UErrorCode status = U_ZERO_ERROR;                            \
+        _##name##Regex = til::ICU::CreateRegex(_##name, 0, &status); \
+        if (U_FAILURE(status))                                       \
+        {                                                            \
+            _invalid##name = true;                                   \
+            _##name##Regex.reset();                                  \
+        }                                                            \
+    }
+
+    DEFINE_VALIDATE_FUNCTION(Name);
+    DEFINE_VALIDATE_FUNCTION(Commandline);
+    DEFINE_VALIDATE_FUNCTION(Source);
+
     bool MatchProfilesEntry::MatchesProfile(const Model::Profile& profile)
     {
-        // We use an optional here instead of a simple bool directly, since there is no
-        // sensible default value for the desired semantics: the first property we want
-        // to match on should always be applied (so one would set "true" as a default),
-        // but if none of the properties are set, the default return value should be false
-        // since this entry type is expected to behave like a positive match/whitelist.
-        //
-        // The easiest way to deal with this neatly is to use an optional, then for any
-        // property to match we consider a null value to be "true", and for the return
-        // value of the function we consider the null value to be "false".
-        auto isMatching = std::optional<bool>{};
-
-        if (!_Name.empty())
+        auto isMatch = [](const til::ICU::unique_uregex& regex, std::wstring_view text) {
+            if (text.empty())
+            {
+                return false;
+            }
+            UErrorCode status = U_ZERO_ERROR;
+            uregex_setText(regex.get(), reinterpret_cast<const UChar*>(text.data()), static_cast<int32_t>(text.size()), &status);
+            const auto match = uregex_matches(regex.get(), 0, &status);
+            return status == U_ZERO_ERROR && match;

```

Comment:
To make it easier on ourselves, we could write our own `unique_uregex` struct. This would allow us to define helper functions on it like `matches(std::wstring_view)` that does all of this under the hood for you.

Not sure if it's worth doing as part of this PR, but if you're feeling eager, I think it may be worth it.

Reply:
FWIW, if we do that, I think it's worth keeping the current approach of not throwing when the regex is invalid (and instead return a bool or `std::optional`, etc.). I feel like throwing would be overkill and just splatter try/catch everywhere we handle user input...